File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
class Solution {

    public boolean isValidBST(TreeNode root) {
        return isValid(root,null,null);
    }
    
    public boolean isValid(TreeNode root, Integer max, Integer min)
    {
        if(root==null)   //Base case: Because it reached the end, it means that it did not encounter false
            return true;

        if( (max!=null && root.val>=max) || (min!=null && root.val<=min) )          //Root has to be greater than the left subtree and least of the right subtree
            return false;

        return isValid(root.left,root.val,min) && isValid(root.right,max,root.val);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool checkValid(TreeNode* node, TreeNode* high, TreeNode* low) {
        if(node == nullptr) {
            return true;
        }

        if(high != nullptr && high->val <= node->val ||
            low != nullptr && low->val >= node->val) {
            return false;
        }
        bool left = checkValid(node->left, node, low);
        bool right = checkValid(node->right, high, node);
        return (left && right);
    }
    bool isValidBST(TreeNode* root) {
        return checkValid(root, nullptr, nullptr);
    }
};